avocado_vt: Move runTest to test phase and improve exceptions

ldoktor opened the pull request:
Avocado now supports status CANCEL, which is basically what SKIP was
being used for in Avocado-vt. Let's move the test execution back to
"runTest" phase and use `self.$status` to map the Avocado-vt exceptions
to Avocado ones, which, unlike exceptions, should be stable.

Fixes: https://github.com/avocado-framework/avocado-vt/issues/934

ldoktor commented:
Oh, actually I need to make this compatible with 36lts...

ldoktor commented:
Tested this with (based on @sathnaga's check):

```
from avocado.core import exceptions
from autotest.client.shared import error
import logging

def run(test, params, env):
    """
    Test command: result unit test
    """
    result_param = params.get("result_param")

    if result_param == "autotest_skip":
        raise error.TestNAError("my skip")
    elif result_param == "autotest_fail":
        raise error.TestFail("my fail")
    elif result_param == "autotest_error":
        raise error.TestError("my error")
    elif result_param == "other exception":
        raise Exception("asefsadf")
    elif 'skip' in result_param:
        raise exceptions.TestSkipError("Test Skip")
    elif 'pass' in result_param:
        logging.info("Test Pass")
        pass
    elif 'fail' in result_param:
        raise exceptions.TestFail("Test Fail")
    elif 'error' in result_param:
        raise exceptions.TestError("Test Error")
    else:
        pass
```
```
variants:
    - check:
        type = check
        start_vm = no
        vms = ''
        main_vm = ''
        variants:
            - skip:
                result_param = 'skip'
            - pass:
                result_param = 'pass'
            - fail:
                result_param = 'fail'
            - error:
                result_param = 'error'
            - autotest_skip:
                result_param = 'autotest_skip'
            - autotest_pass:
                result_param = 'autotest_pass'
            - autotest_fail:
                result_param = 'autotest_fail'
            - autotest_error:
                result_param = 'autotest_error'
            - other_exception:
                result_param = "other exception"

only check
```

using 36lts and master

will-Do commented on avocado_vt/test.py:
```diff
@@ -217,28 +218,48 @@ def setUp(self):
-            raise exceptions.TestError(details)
-        except error.TestFail, details:
-            raise exceptions.TestFail(details)
+        except:  # Old-style exceptions are not inherited from Exception()
```
This will also catch TestSkipError, and lead to a TestError as we raise it in runTest

ldoktor replied:
Well people should not use the `avocado` exceptions directly from `avocado-vt`. Anyway if you think this is needed I can add this exception.

The thing with exceptions is that they are part of the stable API in `autotest`, but in `avocado` they are part of `core` and can (and do) change in time. People should never import from `avocado.core` in tests, they should use `self.error`, `self.fail` and so on. In `avocado-vt` we do the stable mapping from `autotest.client.shared.error` exceptions to `avocado` exceptions so people should use those.

will-Do replied:
I see, so the problem here is that tests in tp-libvirt(also tp-qemu) are replacing ```autotest..error``` with ```avocado.cor.exceptions```, in purpose of getting rid of the dependency of autotest. As you mentioned above, it is wrong usage, but lots of tests already done it.

will-Do commented on avocado_vt/test.py:
```diff
@@ -217,28 +218,48 @@ def setUp(self):
+            if not hasattr(self, "cancel"):     # Old Avocado, skip here
+                if isinstance(self.__status, error.TestNAError):
+                    if hasattr(exceptions, "TestSetupSkip"):
+                        raise exceptions.TestSetupSkip(self.__status)
```
Just curious about this exception, does TestSkipError is not enough?

ldoktor replied:
True, I'm sorry I mistaken the latest changes to Skip exceptions in Avocado. I'll double check and update the PR.

will-Do commented:
BTW, is there any plan about backport the 'Introduce the CANCEL'  patch to 36lts, so we can move on with this new exception?

nasastry commented on avocado_vt/test.py:
```diff
@@ -217,28 +218,48 @@ def setUp(self):
+        elif isinstance(self.__status, error.TestFail):
+            self.fail(self.__status)
+        else:
+            raise self.__status
```
Where the TestErrors will be caught?

ldoktor replied:
Avocado-vt does not use custom handling anymore https://github.com/avocado-framework/avocado-vt/commit/37a315b8577fc8355fdd3b7c5b928f8507206eb2 so the Avocado base class will treat any unknown exceptions as TestError.

ldoktor commented:
@will-Do I added the support for `TestSkipError` (still I don't like it much) and removed the unnecessary `TestSetupSkip` exception. I tested it on `36.0lts 36.3 40.0 44.0 45.0 46.0 47.0` and it worked well.

will-Do commented:
@ldoktor 
Understood, thanks.
Another question, as you mentioned :
> People should never import from avocado.core in tests, they should use self.error, self.fail and so on

Guess lots of people don't know this kinds of caveats, is this included in the doc or somewhere, should we give a warning when import these modules outside avocado?

ldoktor commented:
Not sure whether we directly say this, but we do not use `avocado.core` in our `writing tests` documentation and we separate those APIs in http://avocado-framework.readthedocs.io/en/latest/api/test/avocado.html# vs. http://avocado-framework.readthedocs.io/en/latest/api/core/avocado.core.html

will-Do commented:
Got it, thanks.
BTW, I also test this patch, it works well.

will-Do commented:
Guess we can merge this PR now

clebergnu commented:
@will-Do this was already answered by @ldoktor but to make it extra clear, we document the modules as:

* `avocado.utils.*`(http://avocado-framework.readthedocs.io/en/latest/api/utils/avocado.utils.html): *This is a set of utility APIs that Avocado provides as added value to test writers. It’s suppose to be generic, without any knowledge of Avocado and reusable in different projects.*

* `avocado.test` (http://avocado-framework.readthedocs.io/en/latest/api/test/avocado.html): *This is the bare mininum set of APIs that users should use, and can rely on, while writing tests.*

* `avocado.core.*` (http://avocado-framework.readthedocs.io/en/latest/api/core/avocado.core.html): *Internal APIs that may be of interest to Avocado hackers.*

clebergnu commented:
A reference to the mentioned card:

https://trello.com/c/WLI0qiFo/973-make-avocado-core-test-easier-to-be-subclassed

clebergnu commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Define `avocado.core.test.Test` interface to make it easier to be subclassed](https://trello.com/c/WLI0qiFo/973-define-avocado-core-test-test-interface-to-make-it-easier-to-be-subclassed)

will-Do commented:
@clebergnu 
Thanks for the clarify.

ldoktor commented:
Hello @clebergnu, could you please specify what should I change? You requested change but only stated you dislike the already merged patch.

As for the original patch I disagree. Until now Avocado disallowed `skip` outside setUp, while Avocado-vt allows it at any point. The current Avocado is compatible with the current Avocado-vt, but old Avocado is not. To address that I have to use the existing API of the old Avocado, which is `setUp`, `test` and `tearDown`. So we can re-factor the `run_avocado` to rename those methods to `stage1` `stage2` and `stage3`, but it's just renaming.

Anyway, I'm looking forward to discontinue of 36lts because then we'll simplify things and use only `test` method, which looks a bit better, but for now this is the right way to go as we only have 3 entry points in the standard execution (definitely it's cleaner than defining custom `run_avocado`, which is suggested for removal, viz: https://trello.com/c/W58vhyHR/539-some-avocado-test-methods-and-attributes-are-public )

clebergnu commented on virttest/utils_misc.py:
```diff
@@ -2281,7 +2281,7 @@ def _get_backend_dir(params):
     """
     return os.path.join(data_dir.get_root_dir(), 'backends',
-                        params.get("vm_type"))
+                        params.get("vm_type", ""))
```
A small improvement here is not in the documentation that the base directory can be returned with the empty parameter.

ldoktor replied:
I believe the docstring is correct. You always get the backend directory and if you don't provide the vm_type you get the main backends dir, which could work for no-type tests... (it works)

clebergnu commented:
Even though we disagree on the nuances and meaning of `_get_datadir_backend`, the disagreement is a small issue.  I've tested this, and it's definitely an improvement o the current state.

clebergnu commented:
![](https://github.trello.services/images/mini-trello-icon.png) [BUG: Fix broken test statuses](https://trello.com/c/nbixThTU/967-bug-fix-broken-test-statuses)
